src/Git.hs

```diff
@@ -53,21 +61,43 @@ import Format (format)
 import qualified Configuration as Config
 
 -- A branch is identified by its name.
-newtype Branch = Branch Text deriving (Eq)
+newtype Branch = Branch Text deriving newtype (Show, Eq)
+
+-- A branch identified by its name, pointing at origin.
+newtype RemoteBranch = RemoteBranch Text deriving newtype (Show, Eq)
+
+localBranch :: RemoteBranch -> Branch
+localBranch (RemoteBranch name) = Branch name
 
 -- A commit hash is stored as its hexadecimal representation.
-newtype Sha = Sha Text deriving (Eq)
+newtype Sha = Sha Text deriving newtype (Show, Eq)
+
+newtype RemoteUrl = RemoteUrl Text deriving newtype (Show, Eq)
+
+newtype TagName = TagName Text deriving newtype (Show, Eq)
+
+class RefSpec a where
+  refSpec :: a -> String
+
+instance RefSpec Branch where
+  refSpec (Branch name) = Text.unpack name
 
-newtype RemoteUrl = RemoteUrl Text deriving (Eq)
+instance RefSpec RemoteBranch where
+  refSpec (RemoteBranch name) = "origin/" ++ Text.unpack name
 
-instance Show Branch where
-  show (Branch branch) = Text.unpack branch
+instance RefSpec Sha where
+  refSpec (Sha sha) = Text.unpack $ Text.strip sha
 
-instance Show Sha where
-  show (Sha sha) = Text.unpack sha
+instance RefSpec (Sha, Branch) where
+  refSpec (sha, remote) = refSpec sha ++ ":refs/heads/" ++ refSpec remote
 
-instance Show RemoteUrl where
-  show (RemoteUrl url) = Text.unpack url
+instance RefSpec TagName where
+  refSpec (TagName name) = Text.unpack $ Text.strip name
```

Comment:
You're right, we shouldn't strip here because it's stripped here https://github.com/channable/hoff/blob/86d0ea722b0b42a357ede2f429049e229c3328ad/src/Git.hs#L197